architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+# Data Loader Design For FMS-HF-Tuning
+
+**Deciders(s)**: Sukriti Sharma (sukriti.sharma4@ibm.com), Will Johnson (Will.Johnson@ibm.com) , Abhishek Maurya (maurya.abhishek@ibm.com), Dushyant Behl (dushyantbehl@in.ibm.com), Ashok Pon Kumar (ashokponkumar@in.ibm.com)
+
+**Date (YYYY-MM-DD)**:  2024-03-06
+
+**Obsoletes ADRs**:  NA
+
+**Modified By ADRs**:  NA
+
+**Relevant Issues**: [1]
+
+- [Summary and Objective](#summary-and-objective)
+  - [Motivation](#motivation)
+  - [User Benefit](#user-benefit)
+- [Decision](#decision)
+  - [Alternatives Considered](#alternatives-considered)
+- [Consequences](#consequences)
+- [Detailed Design](#detailed-design)
+
+## Summary and Objective
+
+The reason for motivating dataloader design for fms-hf-tuning is to have a unified interface which supports many type of data formats, streaming and non streaming data, weight based data mixing and many others. Full list we focus on is below - 
+
+1. Data Loading -
+
+    1. Different formats of data → Arrow, Parquet etc.
+
+    1. Different modalities of data -> Images, Audio etc.
+
+    1. Streaming data.
+
+    1. Data Replay
+
+    1. Resume with different number of GPUs
+
+    1. Async data loading
+
+1. Data Preprocessing - 
+
+    1. Custom Attn Masking
+
+    1. Tool Usage
+
+1. Data Mixing - 
+
+    1. Static weights based mixing
+
+### Motivation
+
+### User Benefit
+
+## Decision
+
+### Alternatives Considered
+
+## Consequences
+
+### Advantages
+
+### Impact on performance
+
+## Detailed Design
+
+The input spec which user specifies on how to pass information to such data loader is this
+
+```
+dataloader:
```

Comment:
Its not clear to me if this spec will be the default once this feature is merged in. If so, i urge to consider to maintain also the simpler earlier interface where you only need to pass a data path in. So users who do not need to have so complicated dataloader requirements do not need to go through the trouble to write the spec

Reply:
@fabianlim the idea is to keep the simpler interface while also have a detailed spec which can handle complex use cases and variour other requirements.

Reply:
@dushyantbehl Is my understanding right that we can have a set of inbuilt data configs which people can trigger with say some flag? 

For example, we can have:
--data-config-name pretokenized
--data-config-data-path <path>

So that we don't have to maintain multiple data manipulation paths.

Reply:
I think it would be even better if defaults could be selected automatically for simple file formats and use cases - users enter multimodal JSON and default handler is selected . only for users that want to modify it / have a complex use case, we also expose a way to do that